File: cpp/1579_保证图可完全遍历_hard.cpp
class UnionFind {
private:
    vector<int> parent;
    vector<int> sz;
    int setCount;
public:
    UnionFind(int n): setCount(n), parent(n), sz(n, 1) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int x) {
        return parent[x] == x ? x : parent[x] = find(parent[x]);
    }
    bool united(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y)
            return false;
        parent[y] = x;
        sz[x] += sz[y];
        setCount--;
        return true;
    }
    int size(int x) {
        return sz[find(x)];
    }
    int getSetCount() {
        return setCount;
    }
};
class Solution {
public:
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        auto ufa = UnionFind(n), ufb = UnionFind(n);
        int res = 0;
        for (int i=0; i<edges.size(); i++){
            edges[i][1]--;
            edges[i][2]--;
        }
        for (auto &edge : edges) {
            if (edge[0] == 3) {
                if (!ufa.united(edge[1], edge[2]))
                    res++;
                else
                    ufb.united(edge[1], edge[2]);
            }
        }
        for (auto& edge: edges) {
            if (edge[0] == 1) {
                if (!ufa.united(edge[1], edge[2]))
                    res++;
            }
            else if (edge[0] == 2) {
                if (!ufb.united(edge[1], edge[2]))
                    res++;
            }
        }
        if (ufa.getSetCount() != 1 || ufb.getSetCount() != 1)
            return -1;
        return res;
    }
};

File: python/1579_保证图可完全遍历_hard.py
class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        ufa, ufb = UnionFind(n), UnionFind(n)
        res = 0
        for edge in edges:
            edge[1] -= 1
            edge[2] -= 1
        for t, v, u in edges:
            if t == 3:
                if not ufa.united(v, u):
                    res += 1
                else:
                    ufb.united(v, u)
        for t, v, u in edges:
            if t == 1:
                if not ufa.united(v, u):
                    res += 1
            elif t == 2:
                if not ufb.united(v, u):
                    res += 1
        if ufa.setCount != 1 or ufb.setCount != 1:
            return -1
        return res

class UnionFind():
    def __init__(self, n):
        self.n = n
        self.parent = [i for i in range(n)]
        self.sz = [1]*n
        self.setCount = n
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def united(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.parent[y] = x
        self.sz[x] += self.sz[y]
        self.setCount -= 1
        return True
    def size(self, x):
        return self.sz[self.find(x)]